src/foam/nanos/auth/PreventDuplicateEmailLoginService.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+   package: 'foam.nanos.auth',
+   name: 'PreventDuplicateEmailLoginService',
+   extends: 'foam.nanos.auth.ProxyUserLocatorService',
+   flags: ['java'],
+
+  documentation: 'Prevent users from logging in with duplicate email',
+
+  javaImports: [
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.dao.Sink',
+    'foam.nanos.auth.AuthenticationException',
+    'foam.nanos.auth.User',
+    'foam.nanos.theme.Theme',
+
+    'java.util.List',
+
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'superSpid',
+      documentation: 'Set spid of a theme that is accessible to all users',
+      value: "",
+    }
+  ],
+
+  methods: [
+    {
+      name: 'getUser',
+      documentation: 'Helper logic function to reduce code duplication',
+      type: 'User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'identifier',
+          type: 'String'
+        },
+        {
+          name: 'password',
+          class: 'String'
+        }
+      ],
+      javaCode: `
+        DAO userDAO = (DAO) x.get("localUserDAO");
+
+        Sink sink = new ArraySink();
+        sink = userDAO
+          .where(OR(
+            EQ(User.EMAIL, identifier.toLowerCase()),
+            EQ(User.USER_NAME, identifier)))
+          .select(sink);
+        List list = ((ArraySink) sink).getArray();
+
+        if ( list != null ){
+          if ( list.size() == 0 ) {
+            throw new AuthenticationException("User is not found");
+          } else if ( list.size() > 1 ) {
+            throw new AuthenticationException("Duplicate Email.");
+          }
+        }
```

Comment:
functionally correct but conceptually i dont think so. This is a user locator service not auth, so this decorator needs to return whatever user(s) it can find(perhaps should change return type to array?) And then whatever service calls the locator(auth service in this case) should decide what to do with returned data(throw auth exception) @jlhughes ?

Reply:
As the code is, you could add a limit(2) to make more efficient.

Reply:
This service will be added by a spid specific nspec
spid.preventDupli... 
and should simply return null on invalid scenarios (both cases above)
The idea is that the caller doesn't have to concern itself about the locator service call stack.  It simply requests a user, either it's allowed or it isn't. 
This is login, not user creation or signup.

Reply:
okay, I understand. so if null returned to auth service decorators, the auth service will handle the exception.

Reply:
@jlhughes why does this service need to be added by a spid specific nspec? I believe the two decorator combination we've added is the base-case across all spid when we're talking about the login use case. No matter if a spid allows/doesn't allow duplicate emails, if a duplicate email is used to login, our useLocator service will return null.

Reply:
aren't we preventing users from logging in in case duplicate email is used?

Reply:
yes, we are, but there's a difference between completely eliminating email identifier logins on certain spids and only disallowing email login when there exists more than one user sharing your email.